Unable to collect trace using perfcollect

I'm having difficulty collecting traces when using perfcollect over dotnet-trace. When I run the command:

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/master/scenarios/plaintext.benchmarks.yml --scenario plaintext --profile local --application.framework net5.0 --application.collect true
```

The benchmark successfully runs, but is unable to download the trace file:

```
[03:28:17.138] Running session '659ed0caf93549af82e318035417c373' with description ''
[03:28:17.195] Starting job 'application' ...
[03:28:17.199] Submitted job: http://localhost:5010/jobs/10
[03:28:18.215] Job has been selected by the server ...
[03:28:18.216] Job is now building ... http://localhost:5010/jobs/10/buildlog
[03:28:35.221] Job is running ... http://localhost:5010/jobs/10/output
[03:28:35.226] Starting job 'load' ...
[03:28:35.228] Submitted job: http://localhost:5010/jobs/11
[03:28:37.232] Job has been selected by the server ...
[03:28:37.232] Job is now building ... http://localhost:5010/jobs/11/buildlog
[03:28:44.234] Job is running ... http://localhost:5010/jobs/11/output
[03:29:18.248] Stopping job 'load' ...
[03:29:19.264] Deleting job 'load' ...
[03:29:19.266] Stopping job 'application' ...
[03:29:20.275] Collecting trace file, this can take several seconds
...........................................................................................................................
[03:31:23.322] Downloading trace file application.12-08-15-29-20.trace.zip ...
[03:31:23.339] The trace was not captured on the server: System.Net.Http.HttpRequestException: Response status code does not indicate success: 404 (Not Found).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at Microsoft.Crank.Controller.WebUtils.DownloadFileWithProgressAsync(HttpClient httpClient, String uri, String serverJobUri, String destinationFileName) in /_/src/Microsoft.Crank.Controller/WebUtils.cs:line 56
   at Microsoft.Crank.Controller.JobConnection.DownloadTraceAsync() in /_/src/Microsoft.Crank.Controller/JobConnection.cs:line 665
[03:31:23.339] Deleting job 'application' ...
```

When I look at the logs for crank-agent, I see the following:

```
Hosting environment: Production
Content root path: /home/amd/.dotnet/tools/.store/microsoft.crank.agent/0.1.0-alpha.20608.2/microsoft.crank.agent/0.1.0-alpha.20608.2/tools/netcoreapp3.1/any/
Now listening on: http://[::]:5010
Application started. Press Ctrl+C to shut down.
[03:34:42.710] Created temp directory '/tmp/benchmarks-agent/benchmarks-server-19074/yeucba4e.01d'
[03:34:42.714] Agent ready, waiting for jobs...
[03:34:56.723] Processing job 'application' (1) in state New
[03:34:56.723] Acquiring Job 'application' (1)
[03:34:56.723] Checking requirements...
[03:34:56.724] Downloading dotnet-install to '/tmp/benchmarks-agent/benchmarks-server-19074/wwttg1xw.w1r/dotnet-install.sh'
[03:34:56.727] Downloading https://dot.net/v1/dotnet-install.sh
[03:34:57.474] New -> Initializing
[03:34:58.473] Processing job 'application' (1) in state Waiting
[03:34:58.474] Starting job 'application' (1)
[03:34:58.474] Waiting -> Starting
[03:34:58.474] Created temp directory '/tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy'
[03:34:58.502] Cloning https://github.com/aspnet/benchmarks.git with branch 'master'
[03:34:58.517] [STDERR] Cloning into 'benchmarks'...
[03:35:01.413] Benchmarked Application in /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks
[03:35:01.413] Downloading build tools
[03:35:01.413] Installing dotnet runtimes and sdk
[03:35:01.414] Patching project file with Framework References
[03:35:01.600] Runtime: 5.0.1 (Current)
[03:35:01.600] SDK: 5.0.101 (Current)
[03:35:01.601] ASP.NET: 5.0.1 (Current)
[03:35:01.602] Creating custom global.json
[03:35:01.602] Installing SDK '5.0.101' ...
[03:35:10.912] Installing Runtime '5.0.1' ...
[03:35:11.024] Installing ASP.NET runtime '5.0.1' ...
[03:35:11.211] Working directory: /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks
[03:35:11.211] Command line: /tmp/benchmarks-agent/benchmarks-server-19074/yeucba4e.01d/dotnet publish Benchmarks.csproj -c Release -o /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published /p:MicrosoftNETCoreAppPackageVersion=5.0.1 /p:MicrosoftAspNetCoreAppPackageVersion=5.0.1 /p:BenchmarksNETStandardImplicitPackageVersion=5.0.1 /p:BenchmarksNETCoreAppImplicitPackageVersion=5.0.1 /p:BenchmarksRuntimeFrameworkVersion=5.0.1 /p:BenchmarksTargetFramework=net5.0 /p:BenchmarksAspNetCoreVersion=5.0.1 /p:MicrosoftAspNetCoreAllPackageVersion=5.0.1 /p:NETCoreAppMaximumVersion=99.9 /p:MicrosoftNETCoreApp50PackageVersion=5.0.1 /p:GenerateErrorForMissingTargetingPacks=false /p:MicrosoftNETPlatformLibrary=Microsoft.NETCore.App /p:RestoreNoCache=true --framework net5.0 --self-contained -r linux-x64
[03:35:20.190] Application published successfully in 8978.4165 ms
[03:35:20.192] Patching Benchmarks.runtimeconfig.json
[03:35:20.193] Application is self-contained, skipping runtimeconfig.json
[03:35:20.195] Published size: 113800
[03:35:20.195] Copying crossgen to application folder
[03:35:20.195] Downloading runtime package
[03:35:20.195] Downloading https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/flat2/microsoft.netcore.app.runtime.linux-x64/5.0.1/microsoft.netcore.app.runtime.linux-x64.5.0.1.nupkg
[03:35:20.378] Downloading https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet5/nuget/v3/flat2/microsoft.netcore.app.runtime.linux-x64/5.0.1/microsoft.netcore.app.runtime.linux-x64.5.0.1.nupkg
[03:35:20.507] Downloading https://dotnetfeed.blob.core.windows.net/dotnet-core/flatcontainer/microsoft.netcore.app.runtime.linux-x64/5.0.1/microsoft.netcore.app.runtime.linux-x64.5.0.1.nupkg
[03:35:20.665] Downloading https://api.nuget.org/v3/flatcontainer/microsoft.netcore.app.runtime.linux-x64/5.0.1/microsoft.netcore.app.runtime.linux-x64.5.0.1.nupkg
[03:35:21.932] Copied crossgen to /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published
[03:35:21.946] Invoking executable: /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published/Benchmarks
[03:35:21.946]   Arguments:  --nonInteractive true --scenarios plaintext --urls http://localhost:5000 --server Kestrel --kestrelTransport Sockets --protocol http
[03:35:21.946]   Working directory: /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published
[03:35:21.952] Starting counters
[03:35:21.953] Starting event pipe session
[03:35:21.958] IPC endpoint not available, retrying...
[03:35:21.959] Process started: 20348
[03:35:22.059] Starting event pipe session
[03:35:22.059] IPC endpoint not available, retrying...
[03:35:22.159] Starting event pipe session
[03:35:22.162] Event pipe session started
[03:35:22.239]
[03:35:22.241] ASP.NET Core Benchmarks
[03:35:22.242] -----------------------
[03:35:22.242] Current directory: /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published
[03:35:22.242] AspNetCore location: /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published/Microsoft.AspNetCore.Hosting.Abstractions.dll
[03:35:22.244] AspNetCore version: 5.0.1+fc93e595ceffbb1e3e85532bf454e92a6a80dd6b
[03:35:22.244] .NET Runtime location: /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published/System.Private.CoreLib.dll
[03:35:22.244] .NET Runtime version: 5.0.1-servicing.20575.16+b02e13abab3d6f499751af2ad1ae6333fb4e633d
[03:35:22.245] Environment.ProcessorCount: 256
[03:35:22.288] Event pipe source created
[03:35:22.293] Processing event pipe source (application)...
[03:35:22.327] Scenario configuration found in scenarios.json and/or command line args
[03:35:22.330]
[03:35:22.330] The following scenarios were enabled:
[03:35:22.333]   Plaintext -> /plaintext
[03:35:22.333]
[03:35:22.366] Database: None
[03:35:22.366] ConnectionString: Server=(localdb)\mssqllocaldb;Database=aspnet5-Benchmarks;Trusted_Connection=True;MultipleActiveResultSets=true
[03:35:22.377] Using server Kestrel
[03:35:22.377] Server GC is currently ENABLED
[03:35:22.381] Using Sockets with 16 threads
[03:35:22.411] Hosting environment: Production
[03:35:22.411] Content root path: /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published
[03:35:22.411] Now listening on: http://127.0.0.1:5000
[03:35:22.411] Application started. Press Ctrl+C to shut down.
[03:35:22.411] Ready state detected, application is now running...
[03:35:22.411] Running job 'application' (1)
[03:35:22.411] Starting -> Running
[03:35:22.414] Perfcollect started [20630]
[03:35:23.477] Adding job 'load' (2) to group
[03:35:23.477] Processing job 'application' (1) in state Running
[03:35:23.528] Collection started. Press CTRL+C to stop.
[03:35:24.475] Processing job 'load' (2) in state New
[03:35:24.476] Acquiring Job 'load' (2)
[03:35:24.476] Checking requirements...
[03:35:24.476] New -> Initializing
[03:35:25.477] Processing job 'application' (1) in state Running
[03:35:26.476] Processing job 'load' (2) in state Waiting
[03:35:26.476] Starting job 'load' (2)
[03:35:26.476] Waiting -> Starting
[03:35:26.477] Created temp directory '/tmp/benchmarks-agent/benchmarks-server-19074/uam24wjo.h0h'
[03:35:26.477] Cloning https://github.com/dotnet/crank.git with branch 'master'
[03:35:26.479] [STDERR] Cloning into 'crank'...
[03:35:27.167] Benchmarked Application in /tmp/benchmarks-agent/benchmarks-server-19074/uam24wjo.h0h/crank/src/Microsoft.Crank.Jobs.Wrk
[03:35:27.167] Downloading build tools
[03:35:27.167] Installing dotnet runtimes and sdk
[03:35:27.169] Detected target framework: 'netcoreapp3.1'
[03:35:27.169] Patching project file with Framework References
[03:35:27.530] Runtime: 3.1.10 (Current)
[03:35:27.530] SDK: 3.1.404 (Current)
[03:35:27.530] ASP.NET: 3.1.10 (Current)
[03:35:27.531] Patching existing global.json file
[03:35:27.533] Installing SDK '3.1.404' ...
[03:35:34.438] Installing Runtime '3.1.10' ...
[03:35:34.565] Installing ASP.NET runtime '3.1.10' ...
[03:35:34.685] Working directory: /tmp/benchmarks-agent/benchmarks-server-19074/uam24wjo.h0h/crank/src/Microsoft.Crank.Jobs.Wrk
[03:35:34.685] Command line: /tmp/benchmarks-agent/benchmarks-server-19074/yeucba4e.01d/dotnet publish Microsoft.Crank.Jobs.Wrk.csproj -c Release -o /tmp/benchmarks-agent/benchmarks-server-19074/uam24wjo.h0h/crank/src/Microsoft.Crank.Jobs.Wrk/published /p:MicrosoftNETCoreAppPackageVersion=3.1.10 /p:MicrosoftAspNetCoreAppPackageVersion=3.1.10 /p:BenchmarksNETStandardImplicitPackageVersion=3.1.10 /p:BenchmarksNETCoreAppImplicitPackageVersion=3.1.10 /p:BenchmarksRuntimeFrameworkVersion=3.1.10 /p:BenchmarksTargetFramework=netcoreapp3.1 /p:BenchmarksAspNetCoreVersion=3.1.10 /p:MicrosoftAspNetCoreAllPackageVersion=3.1.10 /p:NETCoreAppMaximumVersion=99.9 /p:MicrosoftNETCoreApp31PackageVersion=3.1.10 /p:MicrosoftNETPlatformLibrary=Microsoft.NETCore.App /p:RestoreNoCache=true --framework netcoreapp3.1 --self-contained -r linux-x64
[03:35:39.674] Application published successfully in 4988.4735 ms
[03:35:39.674] Patching Microsoft.Crank.Jobs.Wrk.runtimeconfig.json
[03:35:39.675] Application is self-contained, skipping runtimeconfig.json
[03:35:39.676] Published size: 76401
[03:35:39.677] Invoking executable: /tmp/benchmarks-agent/benchmarks-server-19074/uam24wjo.h0h/crank/src/Microsoft.Crank.Jobs.Wrk/published/Microsoft.Crank.Jobs.Wrk
[03:35:39.677]   Arguments:  -c 256  http://localhost:5000/plaintext  --latency -d 15s -w 15s -t 32 --header "Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7" --header "Connection: keep-alive"  -s scripts/pipeline.lua -- 16
[03:35:39.677]   Working directory: /tmp/benchmarks-agent/benchmarks-server-19074/uam24wjo.h0h/crank/src/Microsoft.Crank.Jobs.Wrk/published
[03:35:39.678] Running job 'load' (2)
[03:35:39.678] Starting -> Running
[03:35:39.678] Starting counters
[03:35:39.678] Starting event pipe session
[03:35:39.679] IPC endpoint not available, retrying...
[03:35:39.679] Process started: 21514
[03:35:39.777] Starting event pipe session
[03:35:39.779] IPC endpoint not available, retrying...
[03:35:39.877] Starting event pipe session
[03:35:39.877] IPC endpoint not available, retrying...
[03:35:39.946] WRK Client
[03:35:39.946] args: -c 256 http://localhost:5000/plaintext --latency -d 15s -w 15s -t 32 --header Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7 --header Connection: keep-alive -s scripts/pipeline.lua -- 16
[03:35:39.977] Starting event pipe session
[03:35:39.978] Event pipe session started
[03:35:40.038] 82 ms
[03:35:40.049] Downloading wrk from https://aspnetbenchmarks.blob.core.windows.net/tools/wrk to /tmp/.crank/wrk
[03:35:40.078] Event pipe source created
[03:35:40.079] Processing event pipe source (load)...
[03:35:41.476] Processing job 'application' (1) in state Running
[03:35:41.624] > wrk -d 15s -c 256 http://localhost:5000/plaintext --latency -t 32 --header "Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7" --header "Connection: keep-alive" -s scripts/pipeline.lua -- 16
[03:35:42.477] Processing job 'load' (2) in state Running
[03:35:43.477] Processing job 'application' (1) in state Running
[03:35:44.477] Processing job 'load' (2) in state Running
[03:35:45.477] Processing job 'application' (1) in state Running
[03:35:46.481] Processing job 'load' (2) in state Running
[03:35:47.480] Processing job 'application' (1) in state Running
[03:35:48.481] Processing job 'load' (2) in state Running
[03:35:49.480] Processing job 'application' (1) in state Running
[03:35:50.481] Processing job 'load' (2) in state Running
[03:35:51.480] Processing job 'application' (1) in state Running
[03:35:52.481] Processing job 'load' (2) in state Running
[03:35:53.481] Processing job 'application' (1) in state Running
[03:35:54.481] Processing job 'load' (2) in state Running
[03:35:55.480] Processing job 'application' (1) in state Running
[03:35:56.481] Processing job 'load' (2) in state Running
[03:35:56.738] > wrk -d 15s -c 256 http://localhost:5000/plaintext --latency -t 32 --header "Accept: text/plain,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7" --header "Connection: keep-alive" -s scripts/pipeline.lua -- 16
[03:35:56.741] [STDERR] ##ChildProcessId:21856
[03:35:56.741] Tracking child process id: 21856
[03:35:57.480] Processing job 'application' (1) in state Running
[03:35:58.480] Processing job 'load' (2) in state Running
[03:35:59.481] Processing job 'application' (1) in state Running
[03:36:00.480] Processing job 'load' (2) in state Running
[03:36:01.482] Processing job 'application' (1) in state Running
[03:36:02.482] Processing job 'load' (2) in state Running
[03:36:03.484] Processing job 'application' (1) in state Running
[03:36:04.487] Processing job 'load' (2) in state Running
[03:36:05.484] Processing job 'application' (1) in state Running
[03:36:06.484] Processing job 'load' (2) in state Running
[03:36:07.483] Processing job 'application' (1) in state Running
[03:36:08.487] Processing job 'load' (2) in state Running
[03:36:09.489] Processing job 'application' (1) in state Running
[03:36:10.490] Processing job 'load' (2) in state Running
[03:36:11.488] Processing job 'application' (1) in state Running
[03:36:11.849] Running 15s test @ http://localhost:5000/plaintext
[03:36:11.849]   32 threads and 256 connections
[03:36:11.849]   Thread Stats   Avg      Stdev     Max   +/- Stdev

     // Removing these statistics as they are sensitive

[03:36:11.978] Event pipe source stopped (load)
[03:36:11.979] Stopping event pipe session (load)...
[03:36:12.488] Processing job 'load' (2) in state Running
[03:36:12.488] Process has exited (0)
[03:36:12.488] Running -> Stopped
[03:36:13.100] Driver stopping job '2'
[03:36:13.488] Process has exited (0)
[03:36:13.488] Processing job 'application' (1) in state Running
[03:36:14.121] Driver deleting job '2'
[03:36:14.122] Driver stopping job '1'
[03:36:14.488] Processing job 'load' (2) in state Deleting
[03:36:14.488] Process has exited (0)
[03:36:14.488] Deleting job 'load' (2)
[03:36:14.497] Stopping heartbeat (load)... Success!
[03:36:14.497] Stopping counters event pipes for job 'load' (2)
[03:36:14.497] Counters stopped
[03:36:14.497] Process stopped (Deleting)
[03:36:14.498] Deleting directory '/tmp/benchmarks-agent/benchmarks-server-19074/uam24wjo.h0h'
[03:36:14.533] SUCCESS
[03:36:14.533] Deleting -> Deleted
[03:36:15.532] Processing job 'application' (1) in state Stopping
[03:36:15.532] Stopping job 'application' (1)
[03:36:15.532] Stopping heartbeat (application)... Success!
[03:36:15.532] Stopping counters event pipes for job 'application' (1)
[03:36:15.532] Stopping event pipe session (application)...
[03:36:15.673] Event pipe session stopped (application)...
[03:36:15.684] [ERROR] source.Process() -> System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Microsoft.Diagnostics.NETCore.Client.ExposedSocketNetworkStream'.
   at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at FastSerialization.IOStreamStreamReader.Fill(Int32 minimum)
   at FastSerialization.MemoryStreamReader.ReadByte()
   at FastSerialization.Deserializer.FindEndTag(SerializationType type, IFastSerializable objectBeingDeserialized)
   at FastSerialization.Deserializer.ReadObjectDefinition(Tags tag, StreamLabel objectLabel)
   at FastSerialization.Deserializer.ReadObject()
   at Microsoft.Diagnostics.Tracing.EventPipeEventSource.Process()
   at Microsoft.Crank.Agent.Startup.<>c__DisplayClass93_0.<StartCountersAsync>b__2() in /_/src/Microsoft.Crank.Agent/Startup.cs:line 3980
[03:36:15.684] Counters stopped
[03:36:15.684] Invoking SIGINT ...
[03:36:15.686] Application is shutting down...
[03:36:15.884] Process has stopped
[03:36:15.884] Process stopped (Stopped)
[03:36:16.883] Processing job 'load' (2) in state Deleted
[03:36:17.884] Processing job 'application' (1) in state TraceCollecting
[03:36:17.886] Stopping PerfCollect
[03:38:17.887] PerfCollect exceeded allowed time, stopping ...
[03:38:17.887] Waiting for process 20630 to stop ...
[03:38:18.889] PerfCollect process has stopped
[03:38:18.889] Trace collected
[03:38:18.889] TraceCollecting ->  TraceCollected
[03:38:18.889] Stopping counters event pipes for job 'application' (1)
[03:38:18.889] [WARNING] No event source open for job 'application' (1)
[03:38:19.183] Downloading trace for job 1
[03:38:19.183] Sending /tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy/benchmarks/src/Benchmarks/published/benchmarks.trace.zip
[03:38:19.183] Trace file doesn't exist
[03:38:19.196] Driver deleting job '1'
[03:38:19.888] Processing job 'load' (2) in state Deleted
[03:38:20.888] Processing job 'application' (1) in state Deleting
[03:38:20.888] Deleting job 'application' (1)
[03:38:20.888] Deleting directory '/tmp/benchmarks-agent/benchmarks-server-19074/mjxbfhj2.quy'
[03:38:20.957] SUCCESS
[03:38:20.957] Deleting -> Deleted
[03:38:21.956] Processing job 'load' (2) in state Deleted
[03:38:22.956] All jobs in group are finished
```

I'm running crank-agent with sudo, so there shouldn't be any permissions issues. I've also confirmed that there is enough space for the trace to be written.

Some other troubleshooting information:

```
amd@amd-DAYTONA-X0:~/projects/Scratch$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.404
 Commit:    470f6754b3

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-x64
 Base Path:   /usr/share/dotnet/sdk/3.1.404/

Host (useful for support):
  Version: 3.1.10
  Commit:  1721e39439

.NET Core SDKs installed:
  3.1.404 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.10 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.10 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

--- a/src/Microsoft.Crank.Agent/JobContext.cs
+++ b/src/Microsoft.Crank.Agent/JobContext.cs
@@ -33,6 +33,7 @@ namespace Microsoft.Crank.Agent
         public Task EventPipeTask { get; set; }
         public bool EventPipeTerminated { get; set; }
 
+        public EventPipeSession EventPipeSession { get; set; }
         public Task CountersTask { get; set; }
         public TaskCompletionSource<bool> CountersCompletionSource { get; set; }
     }

--- a/src/Microsoft.Crank.Agent/Startup.cs
+++ b/src/Microsoft.Crank.Agent/Startup.cs
@@ -125,6 +125,8 @@ namespace Microsoft.Crank.Agent
         private static ulong eventPipeSessionId = 0;
         private static Task eventPipeTask = null;
         private static bool eventPipeTerminated = false;
+
+        private static EventPipeSession eventPipeSession = null;
         private static Task countersTask = null;
         private static TaskCompletionSource<bool> countersCompletionSource = null;
 
@@ -499,6 +501,7 @@ namespace Microsoft.Crank.Agent
                             eventPipeTask = context.EventPipeTask;
                             eventPipeTerminated = context.EventPipeTerminated;
 
+                            eventPipeSession = context.EventPipeSession;
                             countersTask = context.CountersTask;
                             countersCompletionSource = context.CountersCompletionSource;
 
@@ -1182,10 +1185,10 @@ namespace Microsoft.Crank.Agent
 
                                         if (!countersTask.IsCompleted)
                                         {
-                                            Log.WriteLine("[ERROR] Counters could not be stopped in time");
+                                            Log.WriteLine("[ERROR] Counters could not be stopped in time for job '{job.Service}' ({job.Id})");
                                         }
                                         
-                                        Log.WriteLine($"Counters stopped");
+                                        Log.WriteLine($"Counters stopped for job '{job.Service}' ({job.Id})");
                                     }
                                     else
                                     {
@@ -1417,6 +1420,7 @@ namespace Microsoft.Crank.Agent
                             context.EventPipeTask = eventPipeTask;
                             context.EventPipeTerminated = eventPipeTerminated;
 
+                            context.EventPipeSession = eventPipeSession;
                             context.CountersTask = countersTask;
                             context.CountersCompletionSource = countersCompletionSource;
 
@@ -3746,7 +3750,7 @@ namespace Microsoft.Crank.Agent
             }
 
             // Don't wait for the counters to be ready as it could get stuck and block the agent
-            var counterTask = StartCountersAsync(job);
+            var _ = StartCountersAsync(job);
             
             if (job.MemoryLimitInBytes > 0)
             {
@@ -3830,7 +3834,7 @@ namespace Microsoft.Crank.Agent
                     eventLevel: EventLevel.Verbose)
             );
 
-            EventPipeSession session = null;
+            eventPipeSession = null;
 
             var retries = 0;
             while (retries <= 10)
@@ -3838,7 +3842,7 @@ namespace Microsoft.Crank.Agent
                 try
                 {
                     Log.WriteLine("Starting event pipe session");
-                    session = client.StartEventPipeSession(providerList);
+                    eventPipeSession = client.StartEventPipeSession(providerList);
                     break;
                 }
                 catch (ServerNotAvailableException)
@@ -3874,7 +3878,7 @@ namespace Microsoft.Crank.Agent
             var streamTask = Task.Run(() =>
             {
 
-                var source = new EventPipeEventSource(session.EventStream);
+                var source = new EventPipeEventSource(eventPipeSession.EventStream);
 
                 Log.WriteLine("Event pipe source created");
 
@@ -3885,15 +3889,6 @@ namespace Microsoft.Crank.Agent
                     {
                         // TODO: Catch all event counters automatically
                         // And configure the filterData in the provider
-                        //if (!eventData.EventName.Equals("EventCounters"))
-                        //{
-                        //job.Measurements.Enqueue(new Measurement
-                        //{
-                        //    Timestamp = eventData.TimeStamp,
-                        //    Name = eventData.PayloadByName("name").ToString(),
-                        //    Value = eventData.PayloadByName("value")
-                        //});
-                        //}
 
                         if (eventData.EventName.StartsWith("Measure"))
                         {
@@ -3918,56 +3913,51 @@ namespace Microsoft.Crank.Agent
                             });
                         }
                     }
-                };
 
-                source.Dynamic.All += (TraceEvent eventData) => 
-                {
                     // We only track event counters
-                    if (!eventData.EventName.Equals("EventCounters"))
+                    if (eventData.EventName.Equals("EventCounters"))
                     {
-                        return;
+                        var payloadVal = (IDictionary<string, object>)(eventData.PayloadValue(0));
+                        var payloadFields = (IDictionary<string, object>)(payloadVal["Payload"]);
+
+                        var counterName = payloadFields["Name"].ToString();
+
+                        // Skip value if the provider is unknown
+                        if (!providerNames.Contains(eventData.ProviderName, StringComparer.OrdinalIgnoreCase))
+                        {
+                            return;
+                        }
+
+                        // TODO: optimize by pre-computing a searchable structure
+                        var counter = job.Counters.FirstOrDefault(x => x.Provider.Equals(eventData.ProviderName, StringComparison.OrdinalIgnoreCase) && x.Name.Equals(counterName, StringComparison.OrdinalIgnoreCase));
+
+                        if (counter == null)
+                        {
+                            // The counter is not tracked
+                            return;
+                        }
+
+                        var measurement = new Measurement();
+
+                        measurement.Name = counter.Measurement;
+
+                        switch (payloadFields["CounterType"])
+                        {
+                            case "Sum":
+                                measurement.Value = payloadFields["Increment"];
+                                break;
+                            case "Mean":
+                                measurement.Value = payloadFields["Mean"];
+                                break;
+                            default:
+                                Log.WriteLine($"Unknown CounterType: {payloadFields["CounterType"]}");
+                                break;
+                        }
+
+                        measurement.Timestamp = eventData.TimeStamp;
+
+                        job.Measurements.Enqueue(measurement);
                     }
-
-                    var payloadVal = (IDictionary<string, object>)(eventData.PayloadValue(0));
-                    var payloadFields = (IDictionary<string, object>)(payloadVal["Payload"]);
-
-                    var counterName = payloadFields["Name"].ToString();
-
-                    // Skip value if the provider is unknown
-                    if (!providerNames.Contains(eventData.ProviderName, StringComparer.OrdinalIgnoreCase))
-                    {
-                        return;
-                    }
-
-                    // TODO: optimize by pre-computing a searchable structure
-                    var counter = job.Counters.FirstOrDefault(x => x.Provider.Equals(eventData.ProviderName, StringComparison.OrdinalIgnoreCase) && x.Name.Equals(counterName, StringComparison.OrdinalIgnoreCase));
-
-                    if (counter == null)
-                    {
-                        // The counter is not tracked
-                        return;
-                    }
-
-                    var measurement = new Measurement();
-
-                    measurement.Name = counter.Measurement;
-
-                    switch (payloadFields["CounterType"])
-                    {
-                        case "Sum":
-                            measurement.Value = payloadFields["Increment"];
-                            break;
-                        case "Mean":
-                            measurement.Value = payloadFields["Mean"];
-                            break;
-                        default:
-                            Log.WriteLine($"Unknown CounterType: {payloadFields["CounterType"]}");
-                            break;
-                    }
-
-                    measurement.Timestamp = eventData.TimeStamp;
-
-                    job.Measurements.Enqueue(measurement);                
                 };
 
                 try
@@ -3997,15 +3987,17 @@ namespace Microsoft.Crank.Agent
                 Log.WriteLine($"Stopping event pipe session ({job.Service})...");
 
                 // It also interrupts the source.Process() blocking operation
-                session.Stop();
+                eventPipeSession.Stop();
                 Log.WriteLine($"Event pipe session stopped ({job.Service})...");
             });
 
             countersTask = Task.WhenAll(streamTask, stopTask);
             
-            // The event pipe session needs to be dispose after the source is interrupted
-            session.Dispose();
-            session = null;
+            await countersTask;
+
+            // The event pipe session needs to be disposed after the source is interrupted
+            eventPipeSession.Dispose();
+            eventPipeSession = null;
         }
 
         private static void StartCollection(string workingDirectory, Job job)

--- a/src/Microsoft.Crank.Agent/Startup.cs
+++ b/src/Microsoft.Crank.Agent/Startup.cs
@@ -1130,9 +1130,6 @@ namespace Microsoft.Crank.Agent
                                     Log.WriteLine($"{job.State} ->  TraceCollected");
                                     job.State = JobState.TraceCollected;
                                 }
-
-                                StopCounters();
-
                             }
                             else if (job.State == JobState.Starting)
                             {
@@ -4001,11 +3998,14 @@ namespace Microsoft.Crank.Agent
 
                 // It also interrupts the source.Process() blocking operation
                 session.Stop();
-                session.Dispose();
                 Log.WriteLine($"Event pipe session stopped ({job.Service})...");
             });
 
             countersTask = Task.WhenAll(streamTask, stopTask);
+            
+            // The event pipe session needs to be dispose after the source is interrupted
+            session.Dispose();
+            session = null;
         }
 
         private static void StartCollection(string workingDirectory, Job job)

